qiskit/transpiler/passes/utils/error.py

```diff
@@ -0,0 +1,58 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Error pass to be called when an error happens."""
+
+import logging
+import string
+import warnings
+
+from qiskit.transpiler.basepasses import AnalysisPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class Error(AnalysisPass):
+    """Error pass to be called when an error happens."""
+
+    def __init__(self, msg=None, action='raise'):
+        """Error pass.
+
+        Args:
+            msg (str): Error message, if not provided a generic error will be used
+            action (str): the action to perform. Default: 'raise'. The options are:
+              * 'raise': Raises a `TranspilerError` exception with msg
+              * 'warn': Raises a non-fatal warning with msg
+              * 'log': logs in `logging.getLogger(__name__)`
+        """
+        super().__init__()
+        self.msg = msg
+        if action in ['raise', 'warn', 'log']:
+            self.action = action
+        else:
+            raise TranspilerError('Unknown action: %s' % action)
+
+    def run(self, _):
```

Comment:
why is this a pass? it doesn't even operate on a circuit. An "ErrorPass" sounds weird. Trapping an error and returning it should happen outside the pass manager IMO.

Reply:
we need a way to raise (in a broad sense) an error if dag is bad for some reason. We already have a way to see if a dag is bad, we have AnalysisPasses. Also, we want passes to "fix" those errors. A pattern like this:
```python
pm.append(CheckProperty())
pm.append(TryToFix(), conditional=lambda ps: ps['is_bad'])
pm.append(CheckProperty())
pm.append(Error(), conditional=lambda ps: ps['is_bad'])
```

That pattern is similar to what happens here, when `routing_method='none'`:
```python
pm.append(CheckMap(coupling_map))
pm.append(Error(), condition=lambda property_set: not property_set['is_swap_mapped'])
```

Alternative, we could have `ValidationPass`s like suggested in https://github.com/Qiskit/qiskit-terra/pull/3604 . However, that will duplicate code with the AnalysisPass sometimes. For example, a `ValidationMap` pass (raises when the dag is not mapped) would have practically the same code as `CheckMap`.

To avoid the duplication, `ValidationMap` could require `CheckMap`. In that case, `ValidationMap` will also not operate on the circuit, but the property set. More relevantly, this approach will push the condition to check into the pass (instead of the passmanager `condition`). I think that's a fine solution, but it might break in a lot of validation passes with time.

Reply:
Other possible examples for validation pass:
 - Basis is not satisfiable https://github.com/Qiskit/qiskit-terra/issues/3972
 - Circuit size/depth/gate count fits in backend https://github.com/Qiskit/qiskit-terra/issues/1070
 - Check gates after measure (backend dependant?) https://github.com/Qiskit/qiskit-terra/issues/3289